portfolio/src/main/java/com/google/sps/data/MergeBooks.java

```diff
@@ -0,0 +1,174 @@
+package com.google.sps.data;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.sps.data.Book;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+public class MergeBooks {
+  /**
+   * This method merges two books that have the same title
+   * and ISBN and returns a new book combined from both of them
+   * @param original: the original book
+   * @param addition: the book with additional info you
+   * want to add
+   * The order of the params does not matter
+   */
+  public static Book merge(Book original, Book addition) {
+    validate(original);
+    validate(addition);
+    String isbn = getNonNullString(original.isbn(),
+        addition.isbn()); // One of the ISBNs will be null so you only need to have one valid isbn
+    validateISBN(isbn);
+    Preconditions.checkArgument(
+        original.title() == addition.title(), "The books must be of the same title");
+
+    Book.Builder combined_builder = Book.builder();
+    combined_builder.title(original.title());
+    // Combine lists and Sets
+    combined_builder.categories(combineLists(original.categories(), addition.categories()));
+    combined_builder.genre(combineSets(original.genre(), addition.genre()));
+    combined_builder.authors(combineLists(original.authors(), addition.authors()));
+    ArrayList<String> reviews = combineLists(original.reviews(), addition.reviews());
+    for (String each : reviews) {
+      combined_builder.addReview(each);
+    }
+    // Get non-null strings (value should be present in one book and null in other)
+    combined_builder.isbn(getNonNullString(original.isbn(), addition.isbn()));
+    combined_builder.description(getNonNullString(original.description(), addition.description()));
+    combined_builder.infoLink(getNonNullString(original.infoLink(), addition.infoLink()));
+    combined_builder.thumbnail(getNonNullString(original.thumbnail(), addition.thumbnail()));
+    combined_builder.publishedDate(
+        getNonNullString(original.publishedDate(), addition.publishedDate()));
+    combined_builder.publisher(getNonNullString(original.publisher(), addition.publisher()));
+    combined_builder.maturityRating(
+        getNonNullString(original.maturityRating(), addition.maturityRating()));
+    combined_builder.language(getNonNullString(original.language(), addition.language()));
+    // Integers
+    combined_builder.pageCount(getNonNullInt(original.pageCount(), addition.pageCount()));
+
+    return combined_builder.build();
+  }
```

Comment:
For the lists and sets (33-39) the algorithm combines all unique values. For the strings and ints (41-52), the algorithm finds the value that isn't null and adds it to the book it is building. I made it like this because the most popular use of this method will be combining books created by the BookReader and books created from the BookResponseParser, where the fields will only be legitimate in one of the books (one will have a value and the other will be null).